src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -268,52 +269,88 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
             var projects = (await solutionManager.GetNuGetProjectsAsync()).OfType<IDependencyGraphProject>().ToList();
             var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
```

Comment:
First problem is this call is actually mutating `context` [here](https://github.com/NuGet/NuGet.Client/blob/c090005d05f35725457091e253fc159067fcd300/src/NuGet.Core/NuGet.PackageManagement/Projects/MSBuildNuGetProject.cs#L699).
If we try to fix that part it'll affect many more files and I don't want to go down that path for now.
Also even after I moved it inside parallelization code I don't see any change/improvement when I am experimenting, but only makes code more complicated.

Reply:
how did you test the runtime duration?

Reply:
I just looked the log file from PM UI.

Reply:
I'm not sure that's enough/the best way of approaching it. 

Refer to https://github.com/NuGet/NuGet.Client/pull/3605#pullrequestreview-478075817

* The restore operation is comprised of many different phases, and the solution dg spec population is one of them. 
* In your scenario you are testing installation, which similarly to restore contains many different phases that affect the total time. 

So here's a few tips on how to get more reliable numbers. 

* The best is obviously to rely on a perf lab, you can consider https://github.com/NuGet/Client.Engineering/blob/master/docs/Performance-RPS-Tests.md, but unfortunately I don't think the perf lab has enough variety of test cases to truly quatinfy you scenario. 
* Get a variety of test cases. The test cases in https://github.com/NuGet/NuGet.Client/tree/dev/scripts/perftests/testCases are a good start. You can also include small custom 3-5 project solutions for sanity. 
* Use restore. Restore does *exactly* one GetSolutionREstoreSpec call. Personally I am not sure how many the PM UI does, but we really care about the restore experience more because it happens a lot more frequently than install. 
* Don't use experimental instance, actually install the built VSIX into the VS version that you are testing. 
* Make sure you have a warmup run. After each installation, load a solution, wait for the solution to fully load, run 1 restore.
* Rely on the telemetry reports for the performance, rather than just on total values (in this cases you actually have the data). See: https://github.com/NuGet/NuGet.Client/blob/0bd66a42091bebb5224166330dee29647df8220d/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/RestoreTelemetryEvent.cs#L17, which measures the exact step you are fixing. 
* Run *multiple* restores. Consider the following restores. Load sln, wait for the solution load to complete (one or more restore will have run). Run restore, say 5x times. 
* Observe *only* the step that you are care for. 
* Share the before/after numbers with us!
* Run *multiple* restores for testing purposes.

Reply:
One thing I noticed is everytime I run `telemetry tool` my network VPN connection stop working.
I think it mess up with SSL connection and I tried many different way still loose VPN connection including Teams, Nuget etc. 
Not sure how to how to do it for now.

Reply:
Below are metric change after moving `Project.GetPackageSpecsAndAdditionalMessagesAsync` async call inside parallelization: 
Metric difference on desktop:
https://gist.github.com/erdembayar/5899f1c9a7929831f6bc2cc36e334616
Metric difference on laptop:
https://gist.github.com/erdembayar/1256e1a88d6b1043a4f2ea93eb6b540e

For above change, I had to add more locks inside LegacyPackageReferenceProject.cs, ProjectJsonNuGetProject.cs for context. But still not sure if I checked all thread safety checks because context is passed to more other calls. It's hard to track down check for mutation and prevent from future changes introduce mutation :

[https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs#L125](https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs#L125)
[https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L248](https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L248)
[https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L223](https://github.com/NuGet/NuGet.Client/blob/e97d556a83750bb4f8017e3640bc699d04832657/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L223)

Reply:
@nkolev92 I am not sure if we need to merge this PR, gain is very negligible and not consistent, prone to noise, might introduce deadlocks. Give me your thoughts.